src/o365/spo/commands/page/page-clientsidewebpart-add.ts

```diff
@@ -0,0 +1,483 @@
+				headers: Utils.getRequestHeaders({
+					authorization: `Bearer ${accessToken}`,
+					'content-type': 'application/json;charset=utf-8',
+					accept: 'application/json;odata=nometadata'
+				}),
+				json: true
+			};
+
+			if (this.debug) {
+				cmd.log('Executing web request...');
+				cmd.log(requestOptions);
+				cmd.log('');
+			}
+
+			request
+				.get(requestOptions)
+				.then((res: { value: ClientSidePageComponent[] }) => {
+					if (this.debug) {
+						cmd.log('Response:');
+						cmd.log(res);
+						cmd.log('');
+					}
+
+					const webPartDefinition = res.value.filter((c) => c.Id === webPartId);
+					if (webPartDefinition.length == 0) {
+            reject(new Error(`There is no available WebPart with Id ${webPartId}.`));
+            return;
+					}
+
+					if (this.debug) {
+						cmd.log('WebPart definition:');
+						cmd.log(webPartDefinition);
+						cmd.log('');
+					}
+
+					if (this.verbose) {
+						cmd.log(`Creating instance from definition of WebPart ${webPartId}...`);
+					}
+					const webPart = ClientSideWebpart.fromComponentDef(webPartDefinition[0]);
+					resolve(webPart);
+				});
+		});
+	}
+
+	private _setWebPartPropertiesAndLayout(
+		clientSidePage: ClientSidePage,
+		webPart: ClientSideWebpart,
+		cmd: CommandInstance,
+		args: CommandArgs
+	) {
+		let actualSectionIndex: number | null = args.options.section && args.options.section - 1;
+		// If the section arg is not specified, set to first section
+		if (actualSectionIndex == null) {
+			if (this.debug) {
+				cmd.log(`No section argument specified, The component will be added to default section`);
+			}
+			actualSectionIndex = 0;
+		}
+		// Make sure the section is in the range, other add a new section
+		if (actualSectionIndex >= clientSidePage.sections.length) {
+			throw new Error(`Invalid Section '${args.options.section}'`);
+		}
+
+		// Get the target section
+		const section: CanvasSection = clientSidePage.sections[actualSectionIndex];
+
+		let actualColumnIndex: number | null = args.options.column && args.options.column - 1;
+		// If the column arg is not specified, set to first column
+		if (actualColumnIndex == null) {
+			if (this.debug) {
+				cmd.log(`No column argument specified, The component will be added to default column`);
+			}
+			actualColumnIndex = 0;
+		}
+		// Make sure the column is in the range of the current section
+		if (actualColumnIndex >= section.columns.length) {
+			throw new Error(`Invalid Column '${args.options.column}'`);
+		}
+
+		// Get the target column
+		const column: CanvasColumn = section.columns[actualColumnIndex];
+
+		if (args.options.webPartProperties) {
+			if (this.debug) {
+				cmd.log('WebPart properties: ');
+				cmd.log(args.options.webPartProperties);
+				cmd.log('');
+			}
+
+			try {
+				let properties = JSON.parse(args.options.webPartProperties);
+				webPart.setProperties(properties);
+			} catch (error) {
+				throw new Error('WebPart properties cannot be set');
```

Comment:
we shouldn't be throwing exceptions. Instead, we could move parsing the properties JSON to the `validate` method where we can check if it's a valid object before executing the command.